From 206543a9e9b314fcb60fb5d63005142df15f87bf Mon Sep 17 00:00:00 2001
From: Skye Wanderman-Milne <skyewm@google.com>
Date: Fri, 8 Dec 2023 15:32:31 -0800
Subject: [PATCH] [jax] Make from_dlpack error on non-default layouts.

We're currently working on better layout support in jax, but for now,
jax can't really handle Arrays with non-standard layouts. Raise an
error instead of returning wrong results when the Array is passed to a
computation (!). This can be reverted once jax has better support for
non-default layouts.

Related issues:
https://github.com/google/jax/issues/7657
https://github.com/google/jax/issues/17784

PiperOrigin-RevId: 589258555
---
 xla/python/dlpack.cc     | 15 +++++++++++++++
 xla/python/xla_client.py |  2 +-
 2 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/xla/python/dlpack.cc b/xla/python/dlpack.cc
index 42353e5b61559..ba076e27cbdd3 100644
--- a/xla/python/dlpack.cc
+++ b/xla/python/dlpack.cc
@@ -377,6 +377,21 @@ StatusOr<pybind11::object> DLPackManagedTensorToBuffer(
   Shape shape = ShapeUtil::MakeShapeWithDenseLayout(element_type, dimensions,
                                                     minor_to_major);
 
+  // Raise an error if the resulting PjRtBuffer would have a non-default layout.
+  // TODO(skyewm): we do this because JAX doesn't currently have good support
+  // for non-default layouts, and will return wrong results if a non-default
+  // layout is passed to a computation expecting default layouts. Remove this
+  // special case when non-default layouts are better supported by JAX.
+  TF_ASSIGN_OR_RETURN(Layout default_layout, device->client()->GetDefaultLayout(
+                                                 element_type, dimensions));
+  if (shape.layout() != default_layout) {
+    return Unimplemented(
+        "from_dlpack got array with non-default layout with minor-to-major "
+        "dimensions (%s), expected (%s)",
+        absl::StrJoin(shape.layout().minor_to_major(), ","),
+        absl::StrJoin(default_layout.minor_to_major(), ","));
+  }
+
   std::function<void()> on_delete_callback;
   if (dlmt->deleter) {
     on_delete_callback = [dlmt]() { dlmt->deleter(dlmt); };
diff --git a/xla/python/xla_client.py b/xla/python/xla_client.py
index af5ddcf4ddb9d..158f58cfcf6d5 100644
--- a/xla/python/xla_client.py
+++ b/xla/python/xla_client.py
@@ -48,7 +48,7 @@
 
 # Just an internal arbitrary increasing number to help with backward-compatible
 # changes. In JAX, reference this via jax._src.lib.xla_extension_version.
-_version = 220
+_version = 221
 
 # Version number for MLIR:Python components.
 mlir_api_version = 54
